test/e2e/storage/testsuites/snapshotsecrets.go

```diff
@@ -0,0 +1,253 @@
+			// Ideally we could replace this with https://golang.org/pkg/testing/#T.Cleanup
+
+			// Depending on how far the test executed, cleanup accordingly
+			// Execute in reverse order, similar to defer stack
+			for i := len(cleanupSteps) - 1; i >= 0; i-- {
+				err := tryFunc(cleanupSteps[i])
+				framework.ExpectNoError(err, "while running cleanup steps")
+			}
+
+		}
+		ginkgo.BeforeEach(func() {
+			init()
+		})
+		ginkgo.AfterEach(func() {
+			cleanup()
+		})
+
+		ginkgo.Context("", func() {
+			var (
+				vs *unstructured.Unstructured
+				// vscontent *unstructured.Unstructured
+				// vsc    *unstructured.Unstructured
+				secret *v1.Secret
+			)
+
+			ginkgo.BeforeEach(func() {
+				var sr *SnapshotResource
+				cleanupSteps = append(cleanupSteps, func() {
+					framework.ExpectNoError(sr.CleanupResource())
+				})
+				// create a secret
+				secret = &v1.Secret{
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: f.Namespace.Name,
+						Name:      "snapshot-secret",
+					},
+					Data: map[string][]byte{
+						"secret-data": []byte("secret-value-1"),
+					},
+				}
+
+				if secret, err = cs.CoreV1().Secrets(f.Namespace.Name).Create(context.TODO(), secret, metav1.CreateOptions{}); err != nil {
+					framework.Failf("unable to create test secret %s: %v", secret.Name, err)
+				}
+
+				// Create snapshot using secret
+				parameters := map[string]string{
+					"csi.storage.k8s.io/snapshotter-secret-name":      "snapshot-secret",
+					"csi.storage.k8s.io/snapshotter-secret-namespace": f.Namespace.Name,
+				}
+				sr = CreateSnapshotResource(sDriver, config, pattern, pvc.GetName(), pvc.GetNamespace(), parameters)
+				vs = sr.Vs
+			})
+			ginkgo.It("Use secrets field, create/list and delete snapshots", func() {
+
+				ginkgo.By("list the snapshot")
+				// TODO: Check if list snapshot is supported. Log and continue if not supported.
+				ListSnapshotsWithSecrets(dc, vs.GetNamespace(), vs.GetName(), framework.Poll, framework.SnapshotDeleteTimeout)
+				// Snapshots should be allowed to be deleted even if secrets are deleted.
+				// if secret != nil {
+				// 	framework.Logf("Deleting the secret %q...", secret.Name)
+				// 	err := cs.CoreV1().Secrets(f.Namespace.Name).Delete(context.TODO(), secret.Name, metav1.DeleteOptions{})
+				// 	if err != nil {
+				// 		framework.Logf("Delete secret failed: %v", err)
+				// 	}
+				// }
+				ginkgo.By("Delete the snapshot")
+				DeleteSnapshotWithSecrets(dc, vs.GetNamespace(), vs.GetName(), framework.Poll, framework.SnapshotDeleteTimeout)
+				ginkgo.By("checking if the snapshot is deleted")
+
+			})
+		})
+	})
+}
+
+// ListSnapshotsWithSecrets lists a given volumesnapshot it uses secret provided within storageclass
+func ListSnapshotsWithSecrets(dc dynamic.Interface, ns string, snapshotName string, poll, timeout time.Duration) error {
+	ginkgo.By("listing the snapshot")
+	// TODO: list snapshots using secrets
```

Comment:
Can you add this test?  Secrets for ListSnapshots have different keys from those for CreateSnapshot.

Reply:
See my comments below.  For ListSnapshots, just make sure you add Secrets for ListSnapshots and add a test for Pre-provisioned snapshots.